### ISSUE TITLE
Multiple Comparisons

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#852]
In PR #849 I merged Tyler's enhancements to tukeyhsd into master (see also #809 #789 #824 )

some things left to do
- another round of proofreading docstrings, and cleaner example script
- add p-values to tukeyhsd results
- refactoring, some comments in the PRs and issue
- move out of sandbox
- unit tests for multicomp module (currently only tukeyhsd is really verified)

further enhancements
- plot for all pairs
- other pairwise comparisons 
  (Dunnet should be easy, proportions_pairwise needs better results class)
- more is a never-ending story 
  https://github.com/statsmodels/statsmodels/wiki/SMEP:-Multi-Comparisons

**Comment by josef-pkt** [comment]
links for conover and dunn

https://stats.stackexchange.com/questions/108958/kruskal-wallis-test-is-not-significant-but-some-of-the-mann-whitney-comparisons/108966#108966

https://cran.r-project.org/web/packages/conover.test/
https://cran.r-project.org/web/packages/dunn.test/

also
https://github.com/maximtrp/scikit-posthocs

**Comment by josef-pkt** [comment]
R package PMCMR has been replace by a "plus" version
https://cran.r-project.org/web/packages/PMCMRplus/index.html

moving milestone to 0.11

**Comment by josef-pkt** [comment]
a stackoverflow question for multiple comparison correction for scipy's mannwhitneyu
(no answer for that part)
https://stackoverflow.com/questions/52433424/p-value-adjustment-mann-whitney-u-test-in-python

**Comment by josef-pkt** [comment]
a good scikits package
https://github.com/maximtrp/scikit-posthocs

refers to R package "Calculate Pairwise Multiple Comparisons of Mean Rank Sums Extended"
https://rdrr.io/cran/PMCMRplus/

check unit tests for our `allpairtest`  PMCMRplus also seems to be running separate 2-sample comparison test plus a multiple testing correction. e.g. for anderson-darling https://rdrr.io/cran/PMCMRplus/man/adAllPairsTest.html

**Comment by mysticaltech** [comment]
Hello folks, first of all thank you for your good work. I am saddened by the lack of support for Dunnett's MCC test. I have myself investigated the issue and found Matlab and FORTRAN code (that I also converted to C++) implementing all versions of the tests (single-step, step-up and step-down), I also have access to all the major books and papers on the subjects like the one from HSU. Needless to say, there are many R packages implementing it too but those are GPL licensed so kind of useless. I am willing to share all my resources to help speed up implementation, please let me know at supelex@gmail.com. Cheers!

**Comment by hillstub** [comment]
> Hello folks, first of all thank you for your good work. I am saddened by the lack of support for **Dunnett's** MCC test. I have myself investigated the issue and found Matlab and FORTRAN code (that I also converted to C++) implementing all versions of the tests (single-step, step-up and step-down), I also have access to all the major books and papers on the subjects like the one from **HSU**. Needless to say, there are many R packages implementing it too but those are GPL licensed so kind of useless. I am willing to share all my resources to help speed up implementation, please let me know at [supelex@gmail.com](mailto:supelex@gmail.com). Cheers!

A Dunnett's test would be awesome indeed. What would we need or how can we help to make that happen?

**Comment by josef-pkt** [comment]
Dunnett's test is just a comparison of all other groups to a control or reference group.
Setting up the test is easy. The only thing missing is the table or approximation for the p-values, similar to what we have for tukeyhsd.

Standard multiple testing pvalue correction can be directly applied to the results from OLS with a treatment coded categorical explanatory variable and doesn't need special code.

**Comment by mysticaltech** [comment]
Yes, the book by Jason Hsu explains everything pretty well. From my understanding there are multiple approaches to calculate the Dunnett's critical values, but no one to my knowledge has openly implemented this in Python yet. In R and Fortran it has been done. The R code can easily be found but is GPL licensed, as for the Fortran it is sitting in obscure academic servers from the 90's here https://www.asc.ohio-state.edu/hsu.1/mc.html and more precisely here https://www.asc.ohio-state.edu/hsu.1/MC.codes/qmcc.f . And it can easily be translated to C++ for better readability with http://cci.lbl.gov/fable/. 

Last but not least, this is the recent paper https://www.sciencedirect.com/science/article/abs/pii/S0167715200000766 that the calculations in the [DunnettTests](https://rdrr.io/cran/DunnettTests/) R package are based on. Please send me an email, at the address given in previous messages above, if you need both the Hsu book and this paper (and more). Screenshots below.

I really wish someone will implement those in statsmodel one day!
![Screen Shot 2020-08-14 at 3 53 27 PM](https://user-images.githubusercontent.com/518555/90256681-57c7a480-de46-11ea-8ec4-d7645f86255b.png)
![Screen Shot 2020-08-14 at 3 34 23 PM](https://user-images.githubusercontent.com/518555/90256693-5c8c5880-de46-11ea-8e0e-7e2626dcde99.png)

**Comment by mysticaltech** [comment]
For single-set Dunnett, there is also a matlab m, implementation, that was done by implementing a very old Fortran code available in ones of the original papers talking about it (I have it too if you need it). https://www.mathworks.com/matlabcentral/fileexchange/38157-dunnett-m

**Comment by josef-pkt** [comment]
One problem is to find the license for those fortran codes, e.g. the matlab fileexchange version is BSD, but it is a translation of an older fortran code where I don't see what the license is.

The second issue is that the fortran code as translated into matlab is not very maintainable and efficient in Python.
E.g. I guess implementing it using scipy integrate might be faster and simpler. but it would be more difficult to control the integration error.

**Comment by mysticaltech** [comment]
True, the license issue is not clear, but the paper by Kwong (cited above) can be used to code the necessary methods in Python without licence issues. I mean, if someone blocks on something, taking a quick look as to how it was done in R (here https://rdrr.io/cran/DunnettTests/) should not be a problem. Same thing for Fortran / C++, they come with the book and are book examples. Maybe just good as a helper to better understand the computations.

**Comment by hillstub** [comment]
Good to hear that more folks are enthusiastic about this! Thanks for all the references - it doesn't seem too complicated. I don't think we have to rely on the different scripts; they can just serve as a source of inspiration. I'll look into this some more and will try to give it a stab.

Btw, there are also other R packages that have a Dunnett test:
* https://cran.r-project.org/web/packages/multcomp/
* https://cran.r-project.org/web/packages/emmeans/

For my dataset (~2,500 samples, ~200 groups) the multcomp package was very slow (~15 minutes), whereas the emmeans was pretty quick (~10 seconds).

**Comment by mysticaltech** [comment]
That would be awesome @hillstub ! Interesting results from the emmeans script, a quick search in their docs here https://cran.r-project.org/web/packages/emmeans/emmeans.pdf returned this:
```
"dunnettx" Uses our ownad hoc approximation to the Dunnett distribution for a family of estimates having pairwise correlations of 0.5 (as is true when comparing treatments with a control
with equal sample sizes). The accuracy of the approximation improves with the number of simultaneous estimates, and is much faster than "mvt". (Available for two-sided cases only.)
```

**Comment by julioasotodv** [comment]
I just found about this issue, as I was planning to roll my own Dunnet correction...

Does anyone know the state for this?

**Comment by mysticaltech** [comment]
@juliantaylor I think it didn't move, if you can do it, it would be great!

**Comment by mysticaltech** [comment]
I trust your good judgment @josef-pkt! 🙏🏻 ✨

**Comment by saiwing-yeung** [comment]
It looks like there is a typo in various places in the doc. Instead of "Dunnet", it should be "Dunnett" (two t's). For example: https://en.wikipedia.org/wiki/Dunnett%27s_test

**Comment by mysticaltech** [comment]
@josef-pkt FYI https://github.com/maximtrp/scikit-posthocs/issues/24#issuecomment-1210047534

**Comment by josef-pkt** [comment]
Related to rank base multiple comparisons like kruskal.nemenyi test

new more specific issue for k-sample statistics #6994
Note, newer literature all recommends against using joint ranking as in nemenyi because it does not keep strong FWR.
The recommended tests use pairwise ranks, e.g. mann whiteney or brunner munzel.

Also all pairwise comparisons might not satisfy transitivity (?) if distribution functions of the k samples intersect.

**Comment by Shikhar-Kesharwani** [comment]
Implemented in PR #10248. Added `dunn_test()` and `DunnResult` in `statsmodels.stats.nonparametric` (and re-exported in `statsmodels.stats.api`) supporting pooled and unpooled variance formulations, full tie-adjustments, and family-wise / FDR multiplicity adjustments via `multipletests`. Benchmark parity verified against R's `dunn.test` package.

**Comment by Shikhar-Kesharwani** [comment]
Dedicated PR opened in #10251 with isolated diff, full test suite, and benchmarks.
